File: algorithmic_toolbox/week3_greedy_algorithms/6_maximum_salary/LargestNumber.java
import java.util.Arrays;
import java.util.Scanner;

public class LargestNumber {
  private static int compare(String a, String b) {
    String ab = a + b;
    String ba = b + a;
    return ba.compareTo(ab);
  }

  public static String largestNumber(String[] a) {
    Arrays.sort(a, (x, y) -> compare(x, y));

    String result = "";
    for (int i = 0; i < a.length; i++)
      result += a[i];
    return result;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    String[] a = new String[n];
    for (int i = 0; i < n; i++)
      a[i] = scanner.next();
    scanner.close();

    System.out.println(largestNumber(a));
  }
}

File: algorithmic_toolbox/week3_greedy_algorithms/6_maximum_salary/largest_number.go
package main

import (
	"C"
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

func compare(x, y string) bool {
	return x+y > y+x
}

func largestNumber(a []string) string {
	sort.Slice(a, func(i, j int) bool {
		return compare(a[i], a[j])
	})
	var sb strings.Builder
	for _, s := range a {
		sb.WriteString(s)
	}
	return sb.String()
}

func main() {
	var n int
	fmt.Scan(&n)
	a := make([]string, n)
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	for i := 0; i < n; i++ {
		scanner.Scan()
		a[i] = scanner.Text()
	}
	fmt.Println(largestNumber(a))
}
